backend/inspirehep/pidstore/models.py

```diff
@@ -0,0 +1,139 @@
+    )
+    new_pid_id = db.Column(
+        db.Integer,
+        db.ForeignKey(
+            "pidstore_pid.id", name="fk_inspire_pidstore_redirect_new_pid_id"
+        ),
+        nullable=False,
+        index=True,
+    )
+
+    new_pid = db.relationship(
+        "PersistentIdentifier",
+        backref="redirected_pids",
+        foreign_keys=new_pid_id,
+        innerjoin=True,
+        uselist=False,
+    )
+
+    @classmethod
+    def _fix_existing_redirection(cls, redirection, old_pid, new_pid):
+        if old_pid.status != PIDStatus.REDIRECTED:
+            raise PidStatusBroken(old_pid)
+
+        elif (
+            redirection.new_pid.pid_type == new_pid.pid_type
+            and redirection.new_pid.pid_value == new_pid.pid_value
+        ):
+            LOGGER.info(
+                "Pid already redirected correctly.", old_pid=old_pid, new_pid=new_pid
+            )
+        else:
+            redirection.new_pid = new_pid
+            db.session.add(redirection)
+        return redirection
+
+    @classmethod
+    def _create_new_redirection(cls, old_pid, new_pid):
+        try:
+            with db.session.begin_nested():
+                redirection = cls(original_pid=old_pid, new_pid=new_pid)
+                db.session.add(redirection)
+                old_pid.status = PIDStatus.REDIRECTED
+                db.session.add(old_pid)
+        except IntegrityError as e:
+            raise PIDInvalidAction(e)
+        except SQLAlchemyError as e:
+            LOGGER.exception(
+                "Failed to redirect record", old_pid=old_pid, new_pid=new_pid
+            )
+            raise
+        return redirection
+
+    @classmethod
+    def redirect(cls, old_pid, new_pid):
+        """Redirects pid from old_pid to new_pid"""
+        if old_pid.pid_type != new_pid.pid_type:
+            raise WrongPidTypeRedirection(old_pid, new_pid)
+
+        if (
+            not old_pid.is_registered()
+            and not old_pid.is_deleted()
+            and not old_pid.is_redirected()
+        ) or not new_pid.is_registered():
+            raise WrongRedirectionPidStatus(old_pid, new_pid)
+
+        try:
+            redirection = cls.get(old_pid, reload=True)
+            redirection = cls._fix_existing_redirection(redirection, old_pid, new_pid)
+        except PidRedirectionMissing:
+            redirection = cls._create_new_redirection(old_pid, new_pid)
+
+        LOGGER.info("PID redirected successfully", old_pid=old_pid, new_pid=new_pid)
+        return redirection
+
+    @classmethod
+    def get_redirect(cls, pid):
+        while pid.status == PIDStatus.REDIRECTED:
+            pid = cls.get(pid).new_pid
+        return pid
+
+    @classmethod
+    def get(cls, original_pid, reload=False):
+        redirection = None
+        if reload:
+            redirection = cls.query.filter_by(original_pid=original_pid).one_or_none()
+        elif len(original_pid.redirection) == 1:
+            redirection = original_pid.redirection[0]
```

Comment:
could you explain to me how these two cases are different? AFAIU, in the first case you do the query explicitly, in the second case, you use the ORM to do the query for you. (BTW, how come `redirection` is a list? you're setting `uselist=False` in the `backref`, and the unique constraint on the redirect table guarantees that there can be at most one redirection for a given pid).

Reply:
You can actually check in `integration/pidstore/test_models.py` there was a situation when after removing redirection in session we are trying to get it from pidstore its reusing one from cache. So looks like if we modify redirection and do not commit yet (usually in tests, I cannot find use-case in running service) it's not doing again query but it's using cached value from pid.redirection.

For the second question, I'm not setting `uselist=False` on backref (`PersistantIdentifier` model), but on original_pid and new_pid in this model (InspireRedirect). To set it on PersistentIdentifier I would have to do this in PersistantIdentifier model :/ https://docs.sqlalchemy.org/en/13/orm/relationship_api.html#sqlalchemy.orm.relationship.params.uselist

Reply:
But why does it matter if you get it from the session or from the DB again? that should be an implementation detail.

Reply:
Problem is that it's not taking it from the session but from cache, so it's not updated.

Reply:
I added a small update.